src/vcpkg/binarycaching.cpp

```diff
@@ -2310,137 +1856,393 @@ namespace
+    {
+    }
+
+    ReadOnlyBinaryCache::ReadOnlyBinaryCache(BinaryProviders&& providers) : m_config(std::move(providers)) { }
+
+    void ReadOnlyBinaryCache::fetch(View<InstallPlanAction> actions)
+    {
+        std::vector<const InstallPlanAction*> action_ptrs;
+        std::vector<RestoreResult> restores;
+        std::vector<CacheStatus*> statuses;
+        for (auto&& provider : m_config.read)
+        {
+            action_ptrs.clear();
+            restores.clear();
+            statuses.clear();
+            for (size_t i = 0; i < actions.size(); ++i)
+            {
+                if (actions[i].package_abi())
+                {
+                    CacheStatus& status = m_status[*actions[i].package_abi().get()];
+                    if (status.should_attempt_restore(provider.get()))
+                    {
+                        action_ptrs.push_back(&actions[i]);
+                        restores.push_back(RestoreResult::unavailable);
+                        statuses.push_back(&status);
+                    }
+                }
+            }
+            if (action_ptrs.empty()) continue;
+
+            ElapsedTimer timer;
+            provider->fetch(action_ptrs, restores);
+            size_t num_restored = 0;
+            for (size_t i = 0; i < restores.size(); ++i)
+            {
+                if (restores[i] == RestoreResult::unavailable)
+                {
+                    statuses[i]->mark_unavailable(provider.get());
+                }
+                else
+                {
+                    statuses[i]->mark_restored();
+                    ++num_restored;
+                }
+            }
+            msg::println(msgRestoredPackagesFromVendor,
+                         msg::count = num_restored,
+                         msg::elapsed = timer.elapsed(),
+                         msg::value = provider->vendor());
         }
-    };
+    }
 
-    struct AssetSourcesState
+    bool ReadOnlyBinaryCache::is_restored(const InstallPlanAction& action) const
     {
-        bool cleared = false;
-        bool block_origin = false;
-        std::vector<std::string> url_templates_to_get;
-        std::vector<std::string> azblob_templates_to_put;
-        std::vector<std::string> secrets;
-        Optional<std::string> script;
-
-        void clear()
+        if (auto abi = action.package_abi().get())
         {
-            cleared = true;
-            block_origin = false;
-            url_templates_to_get.clear();
-            azblob_templates_to_put.clear();
-            secrets.clear();
-            script = nullopt;
+            auto it = m_status.find(*abi);
+            if (it != m_status.end()) return it->second.is_restored();
         }
-    };
+        return false;
+    }
 
-    struct AssetSourcesParser : ConfigSegmentsParser
+    std::vector<CacheAvailability> ReadOnlyBinaryCache::precheck(View<InstallPlanAction> actions)
```

Comment:
I think it's now possible to write tests for this, and that's probably worth it now that this function is doing complicated stuff to map between universes.